### PR TITLE
[JUJU-3715] Clean up worker on delete model

### PR DESCRIPTION
**Pull request by SimonRichardson**
~~Requires https://github.com/juju/juju/pull/15579. Will rebase once landed.~~

----
 
The following defines a DBManager. The DBManager allows for more
management of a db, this currently includes deleting a db. This is
so when we delete a model we actually stop the worker for that model.
No other access will then be allowed on that model as we've closed it.

Originally, the service layer would just remove the model uuid from
the model table on the controller, but that's not actually enough. If
you have 100s of models you'll still have 100s of workers still working
away. This actually cleans up the workers by removing them completely.

To do this in a sound fashion, the state no longer deletes the model
from the model list. Instead, the responsibility for that falls back
to the dbaccessor worker. To ensure that we correctly order the deletion
and then the stopping we use the special powers that the dbaccessor
worker has, with accessing the db directly and ensuring that everything
is serialized in the singular loop. That way we don't end up with
any out-of-order events if the state deletes the model uuid and then
the runner stops the worker. Serialization is key here.

The only probably is now if you've already got access to a tracked
db and someone else closes it whilst you've got a lease to it, you'll
end up with a closed db. This should cause the workers to die. We'll
need to verify that as we move forward.

For now, this is the mechanics for sorting that out.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju destroy-model default
```

Ensure everything is cleaned up correctly.

### CODE REVIEW

File: core/database/interface.go [hunk 1]
```diff
@@ -13,6 +13,21 @@ type DBGetter interface {
 	GetDB(namespace string) (TrackedDB, error)
 }
 
+// DBManager describes the ability to create and delete databases.
+type DBManager interface {
```
Comment (nvinuesa) [inline core/database/interface.go:17]: why is this file called `interface.go`? seems too generic and we have `trackeddb.go` in the same package, so maybe this should be `dbgetter.go`?

File: domain/modelmanager/service/service.go [hunk 2]
```diff
@@ -45,6 +52,6 @@ func (s *Service) Delete(ctx context.Context, uuid UUID) error {
 		return errors.Annotatef(err, "validating model uuid %q", uuid)
 	}
 
-	err := s.st.Delete(ctx, uuid)
+	err := s.dbManager.DeleteDB(uuid.String())
```
Comment (manadart) [inline domain/modelmanager/service/service.go:55]: For symmetry with `Create`, I would rather:
- `Delete` accepts `DeleteDB` as a closure.
- The removal from the table is done here.
- The accessor worker deletes the tracker worker upon invocation of the closure as its *only* responsibility.

File: worker/dbaccessor/worker.go [hunk 3]
```diff
@@ -371,6 +403,42 @@ func (w *dbWorker) GetDB(namespace string) (coredatabase.TrackedDB, error) {
 	return tracked.(coredatabase.TrackedDB), nil
 }
 
+// DeleteDB deletes the dqlite-backed database that contains the data for
+// the specified namespace.
+// There are currently a set of limitations on the namespaces that can be
+// deleted:
+//   - It's not possible to delete the controller database.
+//   - It currently doesn't support the actual deletion of the database
```
Comment (manadart) [inline worker/dbaccessor/worker.go:411]: Technically, someone holding a `TrackedDB` for a model that gets deleted can still happily use it until we can actually remove the DB right?

File: worker/dbaccessor/worker_integration_test.go [hunk 4]
```diff
@@ -93,30 +93,79 @@ func (s *integrationSuite) TearDownSuite(c *gc.C) {
...
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(db, gc.NotNil)
 }
 
+func (s *integrationSuite) TestWorkerDeletingControllerDB(c *gc.C) {
+	err := s.dbManager.DeleteDB(coredatabase.ControllerNS)
+	c.Assert(err, gc.ErrorMatches, `.*cannot close controller database`)
+}
+
+func (s *integrationSuite) TestWorkerDeletingUnknownDB(c *gc.C) {
```
Comment (manadart) [inline worker/dbaccessor/worker_integration_test.go:125]: This is not deleting the DB.

**Review by manadart — COMMENTED** [review]
Some initial comments.

File: worker/dbaccessor/worker.go (hunk 3 above)
Comment (SimonRichardson) [reply worker/dbaccessor/worker.go:411]: If they hold a `TrackedDB` and we stop and remove the worker, their db will be closed. All in-play transactions/queries will be closed and should/will cause the watchers et al to shut down.

File: worker/dbaccessor/worker_integration_test.go (hunk 4 above)
Comment (SimonRichardson) [reply worker/dbaccessor/worker_integration_test.go:126]: Officially not, but it should. For now, there is a note about this.

Comment (manadart) [reply worker/dbaccessor/worker_integration_test.go:126]: I mean that you're calling `GetDB` rather than `DeleteDB`.

Comment (SimonRichardson) [reply worker/dbaccessor/worker_integration_test.go:126]: Ah I get you...

**Pull request merged**